jerry-core/ecma/base/ecma-helpers-conversion.c

```diff
@@ -917,227 +919,512 @@ ecma_number_to_int32 (ecma_number_t num) /**< ecma-number */
-  ECMA_NUMBER_CONVERSION_128BIT_INTEGER_RIGHT_SHIFT (fraction_uint128_m1);
+  return fp_value;
+} /* ecma_normalize_encoded_fp */
 
-  ECMA_NUMBER_CONVERSION_128BIT_INTEGER_ADD (fraction_uint128_p1, fraction_uint128);
-  ECMA_NUMBER_CONVERSION_128BIT_INTEGER_RIGHT_SHIFT (fraction_uint128_p1);
-
-  /* While fraction doesn't fit to integer, divide it by 10
-       and simultaneously increment decimal exponent */
-  uint64_t digits_min, digits_max;
+/**
+ * Convert ecma number to encoded_fp
+ *
+ * @return encoded_fp
+ */
+static inline encoded_float_t __attr_always_inline___
+ecma_number_to_encoded_fp (ecma_number_t value) /**< ecma number */
+{
+  encoded_float_t fp;
+  uint64_t *tmp = (uint64_t *) &value;
+  uint64_t u64 = *tmp;
 
-  while (!ECMA_NUMBER_CONVERSION_128BIT_INTEGER_IS_HIGH_BIT_MASK_ZERO (fraction_uint128_m1, 63))
+  if (!(u64 & GRISU3_EXPONENT_MASK))
   {
-    ECMA_NUMBER_CONVERSION_128BIT_INTEGER_DIV_10 (fraction_uint128_m1);
-    decimal_exp_m1++;
+    fp.fract = u64 & GRISU3_FRACTION_MASK;
+    fp.exp = 1 - GRISU3_EXPONENT_BIAS;
   }
-  while (!ECMA_NUMBER_CONVERSION_128BIT_INTEGER_IS_HIGH_BIT_MASK_ZERO (fraction_uint128_p1, 63))
+  else
   {
-    ECMA_NUMBER_CONVERSION_128BIT_INTEGER_DIV_10 (fraction_uint128_p1);
-    decimal_exp_p1++;
+    fp.fract = (u64 & GRISU3_FRACTION_MASK) + GRISU3_IMPLICIT_ONE;
+    fp.exp = (int) ((u64 & GRISU3_EXPONENT_MASK) >> GRISU3_EXPONENT_POSITION) - GRISU3_EXPONENT_BIAS;
   }
 
-  ECMA_NUMBER_CONVERSION_128BIT_INTEGER_ROUND_MIDDLE_AND_LOW_TO_UINT64 (fraction_uint128_m1, digits_min);
-  ECMA_NUMBER_CONVERSION_128BIT_INTEGER_ROUND_MIDDLE_AND_LOW_TO_UINT64 (fraction_uint128_p1, digits_max);
+  return fp;
+} /* ecma_number_to_encoded_fp */
 
-  digits_min++;
+/**
+ * Powers of 10
+ */
+static const uint32_t power_10[] =
+{
+  0u,
+  1u,
+  10u,
+  100u,
+  1000u,
+  10000u,
+  100000u,
+  1000000u,
+  10000000u,
+  100000000u,
+  1000000000u
+};
 
-  if (decimal_exp_m1 < decimal_exp_p1)
-  {
-    JERRY_ASSERT (decimal_exp_m1 == decimal_exp_p1 - 1);
+/**
+ * Get the largest possoble power
+ *
+ * @return power of 10
+ */
+static inline int __attr_always_inline___
+ecma_largest_power (uint32_t n, /**< number */
+                    int n_bits, /**< n_bits */
+                    uint32_t *power) /**< [out] power of 10 */
+{
+  int guess = (((n_bits + 1) * 1233) >> 12) + 1;
 
-    digits_min /= 10;
-    decimal_exp_m1++;
+  if (n < power_10[guess])
+  {
+    --guess; /* We don't have any guarantees that 2^n_bits <= n. */
   }
-  else if (decimal_exp_m1 > decimal_exp_p1)
+
+  *power = power_10[guess];
+
+  return guess;
+} /* ecma_largest_power */
+
+/**
+ * Try all possible numbers with the same leading length
+ * and pick the one that is the closest.
+ *
+ * @return true - if success
+ *         false - otherwise
+ */
+static inline bool __attr_always_inline___
+ecma_grisu_round_weed (char *buffer, /**< [in, out] digits buffer */
```

Comment:
For the sake of precision: this function comes from figure 8 of the paper. Could be noted.